sns/ut/isal_ub.c

```diff
@@ -0,0 +1,407 @@
+                break;
+            }
+        }
+        if (!found) {
+            err_list[i++] = err_id;
+        }
+    } while ( i < failed->b_nob );
+}
+
+static void encode_data(uint32_t k, uint32_t p, uint32_t len,
+                        uint8_t *encode_matrix, uint8_t *g_tbls,
+                        uint8_t **fragments)
+{
+    /* Initialize g_tbls from encode matrix */
+    ec_init_tables(k, p, &encode_matrix[k * k], g_tbls);
+
+    /* Generate EC parity blocks from sources */
+    ec_encode_data(len, k, p, g_tbls, fragments, &fragments[k]);
+}
+
+static int decode_data(uint32_t k, uint32_t p, uint32_t len,
+                       uint8_t *encode_matrix, uint8_t *g_tbls,
+                       uint8_t **fragments,
+                       struct m0_buf *failed, uint8_t **recover_outp)
+{
+    uint32_t i, j, r;
+    uint32_t m = k+p;
+    uint8_t s;
+    uint8_t *temp_matrix = NULL;
+    uint8_t *invert_matrix = NULL;
+    uint8_t *decode_matrix = NULL;
+    uint8_t *recover_srcs[KMAX] = {NULL};
+    uint8_t *err_list = (uint8_t *)failed->b_addr;
+    uint8_t frag_in_err[MMAX];
+    int8_t ret = 0;
+
+    /* Allocate memory for matrices */
+    temp_matrix = m0_alloc(m * k);
+    invert_matrix = m0_alloc(m * k);
+    decode_matrix = m0_alloc(m * k);
+    if ((temp_matrix == NULL) || (invert_matrix == NULL) ||
+        (decode_matrix == NULL)){
+        printf("Error with m0_alloc\n");
+        ret = -1;
+        goto exit;
+    }
+
+    memset(frag_in_err, 0, sizeof(frag_in_err));
+
+    /* Order the fragments in erasure for easier sorting */
+    for (i = 0; i < failed->b_nob; i++) {
+        frag_in_err[err_list[i]] = 1;
+    }
+
+    /* Construct temp_matrix (matrix that encoded remaining frags)
+       by removing erased rows */
+    for (i = 0, r = 0; i < k; i++, r++) {
+        while (frag_in_err[r])
+            r++;
+        for (j = 0; j < k; j++)
+            temp_matrix[k * i + j] = encode_matrix[k * r + j];
+        recover_srcs[i] = fragments[r];
+    }
+
+    /* Invert matrix to get recovery matrix */
+    ret = gf_invert_matrix(temp_matrix, invert_matrix, k);
+    if ( ret != 0)
+        goto exit;
+
+    /* Create decode matrix */
+    for (p = 0; p < failed->b_nob; p++)
+    {
+        /* Get decode matrix with only wanted recovery rows */
+        if (err_list[p] < k)    /* A src err */
+        {
+            for (i = 0; i < k; i++)
+                decode_matrix[k * p + i] =
+                    invert_matrix[k * err_list[p] + i];
+        }
+        /* For non-src (parity) erasures need to multiply
+           encode matrix * invert */
+        else /* A parity err */
+        {
+            for (i = 0; i < k; i++)
+            {
+                s = 0;
+                for (j = 0; j < k; j++)
+                    s ^= gf_mul(invert_matrix[j * k + i],
+                            encode_matrix[k * err_list[p] + j]);
+                decode_matrix[k * p + i] = s;
+            }
+        }
+    }
```

Comment:
Put the opening parenthesis at the same line.

Reply:
Done.